prov/shm/src/smr_ep.c

```diff
@@ -167,13 +167,16 @@ void smr_progress_resp(struct smr_ep *ep)
 	struct smr_resp *resp;
 
 	fastlock_acquire(&ep->region->lock);
-	while (!ofi_cirque_isempty(smr_resp_queue(ep->region))) {
+	fastlock_acquire(&ep->util_ep.tx_cq->cq_lock);
+	while (!ofi_cirque_isempty(smr_resp_queue(ep->region)) &&
+	       !ofi_cirque_isfull(ep->util_ep.tx_cq->cirq)) {
 		resp = ofi_cirque_head(smr_resp_queue(ep->region));
-		if (resp->status)
+		if (resp->status == FI_EBUSY)
 			break;
-		ep->tx_comp(ep, NULL, resp->flags);
+		ep->tx_comp(ep, (void *) resp->msg_id, FI_SEND | resp->flags);
```

Comment:
Are you sure that this is a send completion, versus rma or atomic?

Reply:
For now, yes. When rma and atomics are implemented, no.